core/src/main/java/tc/oc/pgm/shops/menu/ShopMenu.java

```diff
@@ -0,0 +1,202 @@
+      if (col > 7) {
+        col = 1;
+        row++;
+      }
+    }
+  }
+
+  private Category getCategory() {
+    if (category == null) {
+      category = shop.getCategories().get(0);
+    }
+
+    return category;
+  }
+
+  private void setCategory(Category category, int slot) {
+    this.category = category;
+    this.highlightSlot = slot;
+  }
+
+  private ClickableItem[] getCategoryItems() {
+    ClickableItem[] items = new ClickableItem[shop.getCategories().size()];
+    for (int i = 0; i < shop.getCategories().size(); i++) {
+      items[i] = getCategoryItem(shop.getCategories().get(i));
+    }
+    return items;
+  }
+
+  private ClickableItem getCategoryItem(Category category) {
+    return ClickableItem.of(
+        new ItemBuilder()
+            .material(category.getMaterial())
+            .name(ChatColor.GOLD + category.getName())
+            .flags(ItemFlag.values())
+            .build(),
+        c -> {
+          setCategory(category, c.getSlot());
+        });
+  }
+
+  private ClickableItem getMenuSeperator(DyeColor color) {
+    return ClickableItem.empty(
+        new ItemBuilder()
+            .material(Material.STAINED_GLASS_PANE)
+            .color(color)
+            .name(" ")
+            .flags(ItemFlag.values())
+            .build());
+  }
+
+  private ClickableItem getPurchasableItem(Icon icon) {
+    boolean canPurchase = shop.canPurchase(icon, getViewer());
+    Component materialName = text(getMaterial(icon.getCurrency())).color(NamedTextColor.GOLD);
+    NamedTextColor purchaseColor = canPurchase ? NamedTextColor.GREEN : NamedTextColor.RED;
+    Component cost =
+        text()
+            .append(translatable("shop.lore.cost", NamedTextColor.GRAY))
+            .append(text(": ", NamedTextColor.DARK_GRAY))
+            .append(text(icon.getPrice(), purchaseColor))
+            .append(space())
+            .append(materialName)
+            .build();
+    Component click =
+        translatable("shop.lore." + (canPurchase ? "purchase" : "insufficient"), purchaseColor);
+
+    String costLore = TextTranslations.translateLegacy(cost, getBukkit());
+    String clickLore = TextTranslations.translateLegacy(click, getBukkit());
+
+    ItemStack item = icon.getItem().clone();
+    ItemMeta meta = item.getItemMeta();
+    List<String> lore = Lists.newArrayList();
+    if (meta.getLore() != null) {
+      lore.addAll(meta.getLore());
+    }
+    lore.add(costLore);
+    lore.add(clickLore);
+    meta.setLore(lore);
+    meta.addItemFlags(ItemFlag.values());
+    item.setItemMeta(meta);
+
+    return ClickableItem.of(
+        item,
+        c -> {
+          shop.purchase(icon, getViewer());
+          getViewer().getBukkit().updateInventory();
+        });
+  }
+
+  private ClickableItem getPageItem(Player player, int page, boolean next) {
+    return ClickableItem.of(
+        getPageIcon(page, next),
+        c -> {
+          getInventory().open(player, page);
+          this.highlightSlot = 1;
```

Comment:
this looks arbitrary? is there pagination here? i thought the max was 28 items because there is no pagination, only category -> items